Add boolean to GetLabelsForVolume signitature

mvladev opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**: 
The current implementation does not take into account if the cloud providers can actually handle the PVs send to them in `GetLabelsForVolume` method. 

Initializers are only removed, if the volume is supported by the cloud provider and it's not being provisioned.
 
**Which issue(s) this PR fixes** Related to https://github.com/kubernetes/kubernetes/pull/70462#discussion_r229950139

**Special notes for your reviewer**: 

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
Fix cloud-controller-manager's `PersistentVolumeLabel` controller to now update `PersistentVolumes` only when volume type is supported by the cloud provider and volume is not being provisioned.
```

mvladev commented:
/sig cloud-provider

mvladev commented:
/assign @wlan0

andrewsykim commented:
/assign

andrewsykim commented:
@mvladev one thing that PVL controller seems to do that was left out of the admission controller is to [remove initializers](https://github.com/kubernetes/kubernetes/blob/a5789aa8e4dec416f33c633eece4c522050bce64/pkg/controller/cloud/pvlcontroller.go#L257-L274) (even if `len(volLabels) == 0`). Is this something that needs to be accounted for here? Fwiw, the PVL admission controller is deprecated so it's possible some new logic for it was left out.

andrewsykim commented:
cc @dims

mvladev commented:
@andrewsykim the main issue I'm tackling is that PVs which are not for said provider (like trying to create AWS volumes on GKE cluster) are going to be caught and updated. For example `azure` has this [check](https://github.com/kubernetes/kubernetes/blob/0a0eaacd889c825fbbb3a68847b000a54c6e2814/pkg/cloudprovider/providers/azure/azure_managedDiskController.go#L238-L240).

For the cloud-controller-manager there is no way to know if said volume is even supported by the cloud provider - [PVLabeler](https://github.com/kubernetes/kubernetes/blob/0a0eaacd889c825fbbb3a68847b000a54c6e2814/staging/src/k8s.io/cloud-provider/cloud.go#L233-L235) only has the  `GetLabelsForVolume` which only returns `labels` and `errors`

mvladev commented:
More I think about it - the only reasonable way of identifying if PVs are for said provider is to either

- Add new function in the `PVLabeler` interface .
  (like `SupportedVolume(ctx context.Context, pv *v1.PersistentVolume) bool`

  or 

- Change the signature of `GetLabelsForVolume` and add `bool` to the returned values.

the second option will require more changes in every single cloud-provider.

neolit123 commented:
/sig storage
/priority important-longterm

cheftako commented:
/cc @cheftako

justinsb commented:
/ok-to-test

Changing `GetLabelsForVolume` seems reasonable to me, as the problem with two methods is that if we have to make API calls we often end up making two calls instead of one.  Unless you think that there will be a need to "protected" more calls with `SupportedVolume` ?

justinsb commented on pkg/controller/cloud/pvlcontroller.go:
```diff
@@ -187,19 +187,19 @@ func (pvlc *PersistentVolumeLabelController) addLabelsAndAffinity(key string) er
 
 func (pvlc *PersistentVolumeLabelController) addLabelsAndAffinityToVolume(vol *v1.PersistentVolume) error {
-	var volumeLabels map[string]string
 	// Only add labels if the next pending initializer.
```
Pre-existing, but the comment isn't entirely clear...

Commit:
Add boolean to GetLabelsForVolume signitature

Adds a new check to PVLabeler interface which returns true if volume is
supported and it is not being provisioned.

This will prevent the cloud-controller-manager from removing the
finalizer of Persistent volumes which are created for CloudProvider
which is not the currently supported one.

mvladev commented:
Changed implementation to match @justinsb suggestion.

> @mvladev one thing that PVL controller seems to do that was left out of the admission controller is to remove initializers (even if len(volLabels) == 0). Is this something that needs to be accounted for here? Fwiw, the PVL admission controller is deprecated so it's possible some new logic for it was left out.

@andrewsykim the previous initializer flow was flawed - If volume was being provisioned by the cloud provider, then initializers were removed, but [no labels were set](https://github.com/kubernetes/kubernetes/blob/fbc7114734da7e15e2dc4c476d2546b333a93bc2/pkg/controller/cloud/pvlcontroller.go#L207). After provision, labels were no longer set, because volume doesn't have initializers any more and `GetLabelsForVolume` was not called (see [this](https://github.com/kubernetes/kubernetes/blob/fbc7114734da7e15e2dc4c476d2546b333a93bc2/pkg/controller/cloud/pvlcontroller.go#L189) and [this](https://github.com/kubernetes/kubernetes/blob/fbc7114734da7e15e2dc4c476d2546b333a93bc2/pkg/controller/cloud/pvlcontroller.go#L312-L325)).

dims commented:
@andrewsykim @mvladev is this still needed?
